neptune/new/sync/sync.py

```diff
@@ -0,0 +1,233 @@
+                            name, base_path
+                        ),
+                        file=sys.stderr,
+                    )
+
+    def _register_offline_run(
+        self, project: Project, container_type: ContainerType
+    ) -> Tuple[Optional[ApiExperiment], bool]:
+        try:
+            if container_type == ContainerType.RUN:
+                return self._backend.create_run(project.id), True
+            else:
+                # No need for registering project.
+                # Project must've been registered before.
+                return self._backend.get_run(project.id), False
+        except Exception as e:
+            click.echo(
+                "Exception occurred while trying to create a run "
+                "on the Neptune server. Please try again later",
+                file=sys.stderr,
+            )
+            logging.exception(e)
+            return None, False
+
+    @staticmethod
+    def _move_offline_run(
+        base_path: Path, offline_dir: str, server_id: str, server_type: ContainerType
+    ) -> None:
+        online_dir = f"{server_type.value}__{server_id}"
+        # create async directory for run
+        (base_path / ASYNC_DIRECTORY / online_dir).mkdir(parents=True)
+        # mv offline directory inside async one
+        (base_path / OFFLINE_DIRECTORY / offline_dir).rename(
+            base_path / ASYNC_DIRECTORY / online_dir / "exec-0-offline"
+        )
+
+    def register_offline_runs(
+        self, base_path: Path, project: Project, offline_dirs: Iterable[str]
+    ) -> List[ApiExperiment]:
+        result = []
+        for dir_ in offline_dirs:
+            dir_path = base_path / OFFLINE_DIRECTORY / dir_
+            if dir_path.is_dir():
+                run, registered = self._register_offline_run(
+                    project, container_type=ContainerType.RUN
+                )
+                if run:
+                    self._move_offline_run(
+                        base_path=base_path,
+                        offline_dir=dir_,
+                        server_id=run.id,
+                        server_type=run.type,
+                    )
+                    verb = "registered as" if registered else "recognized as"
+                    click.echo(f"Offline run {dir_} {verb} {get_qualified_name(run)}")
+                    result.append(run)
+            else:
+                click.echo(
+                    f"Offline run {dir_} not found on disk.",
+                    err=True,
+                )
+        return result
+
+    def sync_offline_runs(
+        self,
+        base_path: Path,
+        project_name: Optional[QualifiedName],
+        offline_run_ids: Sequence[str],
+    ):
+        if offline_run_ids:
+            project = get_project(project_name, backend=self._backend)
+            if not project:
+                raise CannotSynchronizeOfflineRunsWithoutProject
+            offline_run_dirs = [f"run__{run_id}" for run_id in offline_run_ids]
```

Comment:
why purposefully make this unable to process `model__3e849bc9-6ab3-4c00-87ee-5a56c2a02a3c`?

Reply:
Because it's not part of this issue.

Reply:
What you're saying make 0 sense to me.

In `iterate_experiments` you're using regex to catch all metadata container types. In `sync_selected_registered_runs` you're using `run.type`, in `sync_all_registered_runs` you're also correctly handling all the types.

Here you could do the same with minimal work (you have the container type in path), yet you purposefully decided to:
1. hardcode "run"
2. break syncing offline runs that have no "run__" prefix (i.e. existing runs from previous client versions).

Why? So someone else will have to find all the parts you left broken and fix them in a separate issue?

Reply:
I've changed this code, but kept explicit assumption about `RUN` type.

I didnt' want to support other types here, because it's not a straightforward change.
There's for example case when user manually types UUID of offline run to sync, and now I don't know whether product will agree on changing this to `model_UUID` format, or assume separate input groups for script.
Right now I've simply assumed `UUID` input format for offline runs.
So I don't want to write code matching given UUID with some (maybe) existing patch to determine type, if there's chance that this code will have to be removed.

We've explicitly assumed support of offline mode for runs only for now.

We have separate story for other types, as it requires contact with product and further code refactor.

Additionally supporting other than run types in register_offline_**runs** is odd to me, and at the same time I'd like to merge this branch instead of refactoring this again and again.

Reply:
discussed offline, marking as resolved